### DISCUSSION TITLE
remix-serve should redirect HTTP traffic to HTTPS

### DISCUSSION THREAD

**Discussion by kentcdodds** [remix-run/remix#2915] [Proposals]
I had to do [this](https://github.com/remix-run/remix-jokes/blob/8f786d9d7fa7ea62203e87c1e0bdaa9bda3b28af/app/root.tsx#L25-L46) and it felt like something that should happen for me automagically.

**Comment by sergiodxa** [comment]
Agree, this is something the server should do, not the app

**Comment by tahabesthive** [comment]
```
if(process.env.SSL_FORCE_REDIRECT){
  if(req.headers['x-forwarded-proto'] !== 'https'){ 
    res.redirect('https://' + req.headers.host) 
  }
}
```

**Reply by cliffordfajardo** [reply]
@tahabesthive - is that something you wrote for the Remix App Server or express template?

**Comment by cliffordfajardo** [comment]
I was looking to add HTTPS support in my Remix App Server

The recommended approach would be use the Remix Express template since that exposes express directly [Github discussion](https://github.com/remix-run/remix/discussions/4059#discussioncomment-3468021)
At least until remix supports supports/exposes middleware functions per this discussion ["Middleware before Remix processes request 2857"
](https://github.com/remix-run/remix/discussions/2857)



### Remix App Server Template
For folks using the Remix App server template, perhaps 1 approach would be to create a utility function (ex: `redirectToHttps`) that would take the `request` and handles the redirect at the top of your root loader.

Pros: 
- not needing to eject out of the Remix App Server. For example I was able to setup SSO using the [`remix-auth-microsoft`](https://github.com/juhanakristian/remix-auth-microsoft) package; want to avoid

Cons: 
- you'll need to add manually add `redirectToHttps` function in your loaders/actions manually and hopefully not forget  (maybe this is temporary until #2857 


```ts
// Kents code from post ^^
const upgradeToHttps = (request: Request) => {
  // upgrade people to https automatically

  let url = new URL(request.url);
  let hostname = url.hostname;
  let proto = request.headers.get("X-Forwarded-Proto") ?? url.protocol;

  url.host =
    request.headers.get("X-Forwarded-Host") ??
    request.headers.get("host") ??
    url.host;
  url.protocol = "https:";

  if (proto === "http" && hostname !== "localhost") {
    return redirect(url.toString(), {
      headers: {
        "X-Forwarded-Proto": "https",
      },
    });
  }
}

// call at root loader
export let loader: LoaderFunction = ({ request }) => {
   upgradeToHttps(request)
};
```
